src/irmin-pack/IO.ml

```diff
@@ -210,28 +240,60 @@ module Unix : S = struct
     | false ->
         let raw = raw ~mode ~version:current_version ~generation:0L file in
         v ~offset:0L ~version:current_version ~generation:0L raw
-    | true ->
+    | true -> (
         let x = Unix.openfile file Unix.[ O_EXCL; mode; O_CLOEXEC ] 0o644 in
         let raw = Raw.v x in
         if fresh then (
           let header =
             {
-              Raw.Header.version = current_version;
+              Raw.Header.version = bin_of_version current_version;
               offset = 0L;
               generation = 0L;
             }
           in
           Raw.Header.set raw header;
           v ~offset:0L ~version:current_version ~generation:0L raw)
         else
-          let { Raw.Header.version; offset; generation } = Raw.Header.get raw in
-          assert (version = current_version);
-          v ~offset ~version ~generation raw
+          let version = Raw.Version.get raw in
+          match version_of_bin version with
+          | Some `V1 ->
+              Log.debug (fun l -> l "[%s] file exists in V1" file);
+              if readonly then
+                invalid_arg
+                  "File is in version 1. Open the store in read-write mode to \
+                   migrate it to version 2";
```

Comment:
We should probably return either e.g. ``Error [ `Unsupported_version of version ]`` or a custom exception here so that Tezos can catch it and propose their own migration process.

Also, this seems to keep the implicit migration semantics of opening in read-write to trigger an automatic upgrade. My impression was that we were dropping that in favour of an explicit `migrate` API?

Reply:
> Also, this seems to keep the implicit migration semantics of opening in read-write to trigger an automatic upgrade. My impression was that we were dropping that in favour of an explicit `migrate` API?

Still not sure what is the best API for `migrate`. I added a [commit with the explicit migrate function](https://github.com/mirage/irmin/pull/1015/commits/192c6ed8721ae5acb2eb80dbc05f979f486395fe) but it requires opening the store, calling migrate right away and then using the new handler, returned by the migrate function (as in the [test](https://github.com/mirage/irmin/pull/1015/commits/192c6ed8721ae5acb2eb80dbc05f979f486395fe#diff-5cb474cabaa0cf00ec627057816da328R64)).
We don't support v1 anymore, meaning that if someone attempts to use a v1 store without calling `migrate`, there will be weird error messages, resulting from an unexpected header format.
Another possibility is to fail right away when opening the store in v1, and having the migrate function act as `v`: opens and migrates the store. But to me this is very similar to automatically calling `migrate` in `v` if the store is in v1.

Reply:
> We don't support v1 anymore, meaning that if someone attempts to use a v1 store without calling `migrate`, there will be weird error messages, resulting from an unexpected header format.

My expectation would be that opening a V1 store with V2 implementation would fail in some documented way, not that it would attempt to operate as if it were a V2 store. (And thus `migrate` would take a path rather than an already-opened store.)

> Another possibility is to fail right away when opening the store in v1, and having the migrate function act as `v`: opens and migrates the store. But to me this is very similar to automatically calling `migrate` in `v` if the store is in v1.

I disagree, because the migration process makes the store incompatible with older code and is a destructive change. I don't expect _opening_ a store to make such non-backwards-compatible changes to the data. There is also the fact that "implicit migration" semantics can be built on top of an explicit migrate function (if Tezos decides to go that route), but I think the reverse is not possible: an auto-migrate solution makes it impossible to determine if a store is in the right format without potentially breaking it.

Reply:
As discussed, `v` now fails if the store is opened in V1. An explicit `migrate` function needs to be called to migrate the store to V2, after which the store can be opened with `v` as usual.